pkg/object/cifs.go

```diff
@@ -0,0 +1,455 @@
+	dialer := &net.Dialer{Timeout: 10 * time.Second}
+	address := net.JoinHostPort(c.host, c.port)
+	netConn, err := dialer.Dial("tcp", address)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to %s: %v", address, err)
+	}
+
+	d := &smb2.Dialer{
+		Initiator: &smb2.NTLMInitiator{
+			User:     c.user,
+			Password: c.password,
+		},
+	}
+
+	conn.session, err = d.Dial(netConn)
+	if err != nil {
+		return nil, fmt.Errorf("SMB authentication failed: %v", err)
+	}
+
+	conn.share, err = conn.session.Mount(c.share)
+	if err != nil {
+		_ = conn.session.Logoff()
+		return nil, fmt.Errorf("failed to mount SMB share %s: %v", c.share, err)
+	}
+
+	return conn, nil
+}
+
+// releaseConnection returns a connection to the pool or closes it if there's an error
+func (c *cifsStore) releaseConnection(conn *cifsConn, err error) {
+	if conn == nil {
+		return
+	}
+
+	if err == nil {
+		select {
+		case c.pool <- conn:
+			return
+		default:
+		}
+	}
+
+	// close connection if there's an error or if the pool is full
+	if conn.session != nil {
+		_ = conn.session.Logoff()
+	}
+}
+
+func (c *cifsStore) withConn(f func(*cifsConn) error) error {
+	conn, err := c.getConnection()
+	if err != nil {
+		return err
+	}
+	err = f(conn)
+	c.releaseConnection(conn, err)
+	return err
+}
+
+func (c *cifsStore) Head(key string) (oj Object, err error) {
+	err = c.withConn(func(conn *cifsConn) error {
+		p := c.path(key)
+		fi, err := conn.share.Lstat(p)
+		if err != nil {
+			return err
+		}
+		isSymlink := fi.Mode()&os.ModeSymlink != 0
+		if isSymlink {
+			// SMB doesn't fully support symlinks like POSIX, but we'll try our best
+			fi, err = conn.share.Stat(p)
+			if err != nil {
+				return err
+			}
+		}
+		oj = c.fileInfo(key, fi, isSymlink)
+		return nil
+	})
+	return oj, err
+}
+
+func (c *cifsStore) Get(key string, off, limit int64, getters ...AttrGetter) (io.ReadCloser, error) {
+	var readCloser io.ReadCloser
+	err := c.withConn(func(conn *cifsConn) error {
+		p := c.path(key)
+		f, err := conn.share.Open(p)
+		if err != nil {
+			return err
+		}
+		finfo, err := f.Stat()
+		if err != nil {
+			_ = f.Close()
+			return err
+		}
+		if finfo.IsDir() || off > finfo.Size() {
+			_ = f.Close()
+			readCloser = io.NopCloser(bytes.NewBuffer([]byte{}))
+			return nil
+		}
+		if limit > 0 {
+			readCloser = &SectionReaderCloser{
+				SectionReader: io.NewSectionReader(f, off, limit),
+				Closer:        f,
+			}
+			return nil
+		}
+		readCloser = f
+		return nil
+	})
+	return readCloser, err
+}
+
+func (c *cifsStore) Put(key string, in io.Reader, getters ...AttrGetter) (err error) {
+	return c.withConn(func(conn *cifsConn) error {
+		p := c.path(key)
+		if strings.HasSuffix(p, dirSuffix) {
+			// perm will not take effect, is not used
+			// ref: https://github.com/hirochachacha/go-smb2/blob/c8e61c7a5fa7bcd1143359f071f9425a9f4dda3f/client.go#L341-L370
+			return conn.share.MkdirAll(p, 0755)
+		}
+
+		// Ensure parent directories exist
+		dirPath := path.Dir(p)
```

Comment:
Firstly assume that it already exists, and create it only when an error occurs for better performance.